src/main/java/org/opensearch/neuralsearch/util/ProcessorDocumentUtils.java

```diff
@@ -0,0 +1,115 @@
+                    Locale.getDefault(),
+                    "[%s] configuration doesn't match actual value type, configuration type is: %s, actual value type is: %s",
+                    sourceKey,
+                    fieldMap.getClass().getName(),
+                    sourceValue.getClass().getName()
+                )
+            );
+        }
+        // next level validation, only validate the keys in configuration.
+        ((Map<String, Object>) fieldMap).forEach((key, nextFieldMap) -> {
+            Object nextSourceValue = sourceValue.get(key);
+            if (nextSourceValue != null) {
+                if (nextSourceValue instanceof List) {
+                    validateListTypeValue(key, (List) nextSourceValue, fieldMap, depth + 1, maxDepth, allowEmpty);
+                } else if (nextSourceValue instanceof Map) {
+                    validateMapTypeValue(key, (Map<String, Object>) nextSourceValue, nextFieldMap, depth + 1, maxDepth, allowEmpty);
+                } else if (!(nextSourceValue instanceof String)) {
+                    throw new IllegalArgumentException("map type field [" + key + "] is neither string nor nested type, cannot process it");
+                } else if (!allowEmpty && StringUtils.isBlank((String) nextSourceValue)) {
+                    throw new IllegalArgumentException("map type field [" + key + "] has empty string value, cannot process it");
+                }
+            }
+        });
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private static void validateListTypeValue(
+        String sourceKey,
+        List sourceValue,
+        Object fieldMap,
+        int depth,
+        long maxDepth,
+        boolean allowEmpty
+    ) {
+        validateDepth(sourceKey, depth, maxDepth);
+        if (sourceValue == null || sourceValue.isEmpty()) return;
+        Object firstNonNullElement = sourceValue.stream().filter(Objects::nonNull).findFirst().orElse(null);
+        if (firstNonNullElement == null) return;
+        for (Object element : sourceValue) {
+            if (firstNonNullElement instanceof List) { // nested list case.
+                validateListTypeValue(sourceKey, (List) element, fieldMap, depth + 1, maxDepth, allowEmpty);
+            } else if (firstNonNullElement instanceof Map) {
+                validateMapTypeValue(
+                    sourceKey,
+                    (Map<String, Object>) element,
+                    ((Map) fieldMap).get(sourceKey),
+                    depth + 1,
+                    maxDepth,
+                    allowEmpty
+                );
+            } else if (!(firstNonNullElement instanceof String)) {
+                throw new IllegalArgumentException("list type field [" + sourceKey + "] has non string value, cannot process it");
+            } else {
+                if (element == null) {
+                    throw new IllegalArgumentException("list type field [" + sourceKey + "] has null, cannot process it");
```

Comment:
when you check `if (firstNonNullElement instanceof List)` and `else if (firstNonNullElement instanceof Map)`, `element` could also be null, but no exception thrown, the logic seems inconsistent. Why we check both firstNonNullElement and element, it's a bit confusing

Reply:
Make sense, I'll change this part to follow a consistent principle: for list type, we don't allow null element in it.